common/src/collision.rs

```diff
@@ -0,0 +1,129 @@
+use crate::math;
+
+/// Performs shape casting (swept collision query) against the voxels in the `DualGraph` for a spherical colider
+/// Temporarily performs ray casting instead for testing purposes, as if the sphere has radius 0.
+///
+/// The `start_node_transform` parameter determines which coordinate system the `ray` parameter and any resulting hit
+/// normals are given in. Specifically, the `start_node_transform` matrix converts this coordinate system to the coordinate
+/// system of `start_chunk`'s node.
+///
+/// The `tanh_distance` is the hyperbolic tangent of the distance along the ray to check for hits.
+pub fn shape_cast(
+    start_node_transform: na::Matrix4<f32>,
+    ray: &Ray,
+    tanh_distance: f32,
+) -> RayEndpoint {
+    // A collision check is assumed to be a miss until a collision is found.
+    // This `endpoint` variable gets updated over time before being returned.
+    let mut endpoint = RayEndpoint {
+        tanh_distance,
+        hit: None,
+    };
+
+    let node_ray = start_node_transform * ray;
+
+    // As a temporary placeholder implementation, check for collision against the z=0 plane going in the -z direction.
+
+    const EPSILON: f32 = 1e-4;
+    let found_collision = if node_ray.direction.z >= 0.0 {
+        // Going the wrong way. Don't report a collision.
+        false
+    } else if (-EPSILON..=0.0).contains(&node_ray.position.z) {
+        // Extra logic to ensure precision issues don't allow a collider to clip through a surface
+        endpoint.tanh_distance = 0.0;
+        true
+    } else {
+        // Compute needed tanh_distance to reach z=0 plane.
+        let tanh_distance = -node_ray.position.z / node_ray.direction.z;
+        if tanh_distance >= 0.0 && tanh_distance < endpoint.tanh_distance {
+            // Report a collision if the needed tanh_distance is in the right range
+            endpoint.tanh_distance = tanh_distance;
+            true
+        } else {
+            false
+        }
+    };
+
+    if found_collision {
+        endpoint.hit = Some(RayHit {
+            normal: math::mtranspose(&start_node_transform) * na::Vector4::z(),
+        });
+    }
+
+    endpoint
+}
+
+/// Where a cast ray ended, and all information about the hit at the end if such a hit occurred.
+#[derive(Debug)]
+pub struct RayEndpoint {
```

Comment:
s/`Ray`/`Cast`/ for consistency

Reply:
I had left it as `Ray` for consistency with the `Ray` struct, since this class can act as a companion to it. Still happy to rename, but I would like your confirmation.

Reply:
I think a more generic name makes sense. You can cast rays and spheres and other things, and it's weird to use ray-oriented nomenclature to refer to a sphere-cast result which isn't, in fact, on the ray.

Reply:
That's fair. I'll use the term `Cast` for everything except the `Ray` struct.